### ISSUE TITLE
Correct graph edge for Pod to Pod communication using destination_workload

### ISSUE THREAD

**Issue by MugenTwo** [kiali/kiali#4488]
As suggested by @jmazzitelli in the following discussion: https://github.com/kiali/kiali/discussions/4454 I'm creating this feature request ticket.

### **Background**
I have a microservice architecture in which my Spring Boot Admin Server's pod to pod communication is being represented as PassthroughCluster Node even though **destination_workload** is actually a known entity/app inside the system. 

This makes it so that the traffic is hard to understand because the PassthroughCluster adds a lot of **noise** in the graph, it makes it difficult to do **traffic classification** as I can't identify which traffic is actually an external traffic and which traffic is actually coming from the Spring Boot Admin Server.

### **Use cases**
- Classify traffic of "Spring Boot Admin" type of services which does pod to pod communication for healthchecks, resource monitoring, ...
- Classify traffic for client-side load balancing (e.g. Ribbon). -> more important than dealing with Spring Boot Admin traffic.

### **Suggested behaviour**
Draw and edge between app nodes to represent pod to pod communication with a **new type of edge (maybe different color) by taking into account the destination_workload** instead of of the destination_service_name if the former is a known name and the latter is a PassthroughCluster.

**Comment by jmazzitelli** [comment]
The edges still represent the same kind of data (http, or tcp, or gcp) so this would not warrant a different edge type. Edges are not different based on the endpoint node types.

Because the target node is a workload and not an app (represented with a square node) or service (represented with a triangle node), the graph should simply draw a circle node which represents "workload" as per the graph legend.

**Comment by MugenTwo** [comment]
Agreed on the part that this should be pointing on the correct workload. This is definitely the most important part.

I suggested having a new type -> new color of edge to easily classify the pod to pod communication from the other types of communication... similar to how TCP is being classified with a blue color.

**Comment by jmazzitelli** [comment]
> I suggested having a new type -> new color of edge to easily classify the pod to pod communication from the other types of communication... similar to how TCP is being classified with a blue color.

Yeah, that's the problem. From a UX perspective, you are mixing semantics. Today color/edge type denotes the different protocols (http vs. tcp) flowing over the edge. Suggesting changing color/edge type based on the destination node makes it confusing because now the semantics are a combination of protocol and node type. Think about it this way - what if you have 2 of these specific pod communications - one is sending HTTP traffic to the pod, one is sending TCP traffic to the pod. What color/edge type do you make those edges? Things become ambiguous and confusing to the user.

So colors/edge types wouldn't change for this - we would change the destination node type because that is what is different.

**Comment by MugenTwo** [comment]
Would this sort out the noise in both the App graph? I definitely see that the Workload graph will be sorted out with this. 
But I think the App group uses another criter and the **noise information** will remain on the App graph.

**Comment by jmazzitelli** [comment]
I think when @jshaughn or @nrfox look at this (one of those will most likely be the ones working on this) they will know better. I'm pretty sure just introducing workload nodes for this use-case in the app and versioned-app graphs will solve the problem. But they would know better when they start looking at it.

**Comment by nrfox** [comment]
I agree with @jmazzitelli that having separate edges for workload --> workload communication could make the graph more convoluted given the way that edges/nodes currently represent mesh traffic.

It seems like Kiali is correctly representing what is happening from the perspective of your mesh. Traffic is still going through the sidecar even if that traffic is being sent to the pods directly. From the sidecar, since the request doesn't match a listener it gets sent to the PassthroughCluster which forwards it on to the application. Kiali does not currently have a graph view that filters out all of the service information and just shows the source/destination workloads on the graph. Having this kind of view might be helpful for use cases like this but it would also potentially hide a lot of important routing information.

If traffic is going through the PassthroughCluster, more than likely a service matching the request does not exist in the Istio service registry. For health checks, perhaps these could go through the app's kube Service instead of talking directly to the pods? The client side load balancing might require creating some mesh internal ServiceEntries to get those services into the istio Service Registry. Maybe ribbon has support for this already? I'm just thinking of ways that you can possibly prevent your traffic from being routed to the PassthroughCluster. If Kiali changed the way it represented the PassthroughCluster traffic it might also help with your use case/graph.

Kiali could represent the PassthroughCluster as a service node per app instead of a single global service object. In your graph, each app would have it's own PastthroughCluster service node and it would look something like `Spring boot admin --> PassthroughCluster (MyApp) --> MyApp Workload`. That would help cut down on the noise in the graph while still accurately representing the traffic flow.

In summary, a couple of questions to answer:
1. Should Kiali have a separate graph view that excludes all service traffic and just shows destination/source workload traffic?
2. Should PassthroughCluster be represented as a separate service node in each app rather than a single global service node?

@MugenTwo I think you should also look at trying to configure your mesh so that traffic isn't being routed to the PassthroughCluster if possible.

**Comment by jshaughn** [comment]
How do things look if you disable the Show Service Nodes Display option?  Or alternatively, use a `svc=PassthroughCluster` Hide expression?

**Comment by jmazzitelli** [comment]
This isn't going to be solved by using the display options. Remember the case here - the workload name is known, the service name is not (because there IS no service - its directly going to a pod). The graph needs to show a "workload" node here - there is no service at all here AFAIU.

**Comment by nrfox** [comment]
> The graph needs to show a "workload" node here - there is no service at all here AFAIU.

I think there _is_ a "service", the service is just the PassthroughCluster since that's the default for requests that don't match anything.

> How do things look if you disable the Show Service Nodes Display option?

This sounds very promising if it just shows workload --> workload traffic and filters out the PassthroughCluster service node. I was wrong about Kiali not having an option to display this on the graph :)

**Comment by jshaughn** [comment]
I agree with @nrfox , there is a service, basically the default service being `PassthroughCluster`.  As mentioned earlier, Istio confirmed that the telemetry is as-expected, with destination_service_name="PassthroughCluster".   Having said all of that, I do agree that we want to provide some sort of reasonable visualization if at all possible.  Let's see what @MugenTwo says in response to the above suggestions, and go from there.

**Comment by MugenTwo** [comment]
I'm doing a couple more of tests before coming up with any conclusions. Will reply as soon when I reach a conclusion.

**Comment by MugenTwo** [comment]
### Using svc=PassthroughCluster:
My conclusion is that It hides all the PassthrougClusters and that might not be ideal because it could potentially hide the real PassthroughClusters.
### Disabling the "Service Nodes" Display option:
Similar to the above, it hides all the Service related information which is not good when debugging.

### Suggested options:
1. Users should be able to exclusively and completely filter out PassthroughCluster with known destination_workload.

2. Do what @nrfox suggested here:
> Should PassthroughCluster be represented as a separate service node in each app rather than a single global service node?

But instead of each of them being drawn as a traditional PassthroughCluster, maybe we can use another icon so that they are distinguishable from the other PassthroughClusters. Plus, we can also add a way to filter out these special PassthroughClusters.

**On a side note:**
I've done some testing through CURL and It would seem that whenever there is a pod A to pod B communication, Kiali draws an "edge from pod A (workload) going to the PassthroughCluster which then goes to the Service of pod B" and "another edge from pod A (workload) going directly to the Service of pod B".  My theory is that the former is being drawn based on the destination_service_name and the latter is being drawn using the destination_workload, but I might be mistaken here.

**Comment by jshaughn** [comment]
> `svc=PassthroughCluster` My conclusion is that It hides all the PassthrougClusters and that might not be ideal because it could potentially hide the real PassthroughClusters.

It's true that this would also hide PassthroughCluster when it's used for requests to external hosts (i.e. when PassthroughCluster is a terminal node in the graph). So, I agree that's not ideal.   But, I would not call these the "real" uses of PassthroughCluster, the use case under discussion is no less real, it's just that PassthroughCluster is the default handler for unknown services, when no in-mesh service or serviceEntry matches the request.

As an aside, allowing PassthroughCluster is not usually recommended for production systems, it means `meshConfig.outboundTrafficPolicy.mode` is set to ALLOW_ANY (the default).  I think in general it is recommended to use REGISTRY_ONLY.  See [here](https://istio.io/latest/docs/tasks/traffic-management/egress/egress-control/#envoy-passthrough-to-external-services) for more.

> `Disable "Service Nodes"` , ...it hides all the Service related information which is not good when debugging

I agree, this is a hammer approach, it's good when you want to see things this way, but it's not a good workaround for avoiding a single service node.

> Suggested options:
> Users should be able to exclusively and completely filter out PassthroughCluster with known destination_workload.

This is a fairly specific situation and I'd prefer not to add an explicit filter/option, the number of options is already fairly large.

> Do what @nrfox suggested here (app-specific PassthroughCluster nodes)

This is an interesting idea and would probably add more organization to the graph, but it wouldn't reduce the number of edges.  And it also doesn't really reflect reality, as the PassthroughCluster is not distributed in that way.

My initial thought is that Kiali should just omit the intermediate PassthroughCluster edges.  Meaning, the graph would only show PassthroughCluster when it is a terminal node.  That would result in direct edges from the source workload to the destination workload.  These edges would have the same look as other edges, they would be protocol-specific.

This would reduce the graph verbosity at the expense of being explicit about the PassthroughCluster.  It would be implicit that Istio is routing through the default handler, and that may be OK because as a user it's something of an implementation detail.

@MugenTwo, @nrfox, @jmazzitelli thoughts?

**Comment by jmazzitelli** [comment]
My main question is: "Is Kiali visualizing all the telemetry that Istio stored in Prometheus?" e.g. "Is there a bug here in the graph generation? Or is there just something that we want to show _Better_?". If it is the latter, then I'm less concerned how we solve this issue.

We also need to ask: "What does it mean to see PassthroughCluster in the telemetry?" If this means there is no traffic going through a sidecar (i.e. the traffic is circumventing the Istio traffic configuration) we do need to show SOMETHING here to notify (warn?) the user that this traffic going into a mesh workload is not going through "the mesh". For example, as @jshaughn said, "allowing PassthroughCluster ...means meshConfig.outboundTrafficPolicy.mode is set to ALLOW_ANY" - this is important to know. Hiding PassthroughCluster (and thus not making any indication the traffic policy may be different than what the mesh policy is) would be problematic - the user would have no indication the traffic flowing in the mesh is "outside the mesh" (there would be no indication the traffic is circumventing istio traffic config even though that traffic is entering a workload that is part of the mesh).

I don't have an answer to the solution here - I would just want to point out that if there is traffic that is going into a mesh workload but that traffic is not bound to the traffic policies defined by the mesh, that needs to be visualized somehow.

**Comment by jshaughn** [comment]
> Is Kiali visualizing all the telemetry that Istio stored in Prometheus?

Well, I'd say yes, we're showing the Istio telemetry, but for this use-case the visualization can be too verbose.  So, no, I don't think this is a bug but rather a request for a *better* visualization.

> ... we do need to show SOMETHING here...

Well this is the issue, the request is intended to be handled out-of-mesh and that's why PassthroughCluster is the handler.  If a ServiceEntry were defined I think everything would be fine (note - ServiceEntry would be required if ALLOW_ANY were changed to REGISTRY_ONLY).  It's not really a misconfiguration, and so I'm not sure a warning is appropriate.  As I mentioned above, omitting intermediate PassthroughCluster nodes would make it implicit, but I think it may be required to make the graph more usable in this scenario.

**Comment by nrfox** [comment]
> Well, I'd say yes, we're showing the Istio telemetry, but for this use-case the visualization can be too verbose. So, no, I don't think this is a bug but rather a request for a better visualization.

+1 the visualization seems correct in that the traffic is going to the PassthroughCluster because there are no matching services in the registry for the request. But all of the traffic going into and out from a single PassthroughCluster service node on the graph creates a lot of noise.

> "What does it mean to see PassthroughCluster in the telemetry?" 

As @MugenTwo mentioned, there's at least two different scenarios for this:
1. The destination_workload is unknown. Meaning that the destination is something outside of the mesh possibly a request to an external service like `google.com`
2. The destination_workload is known. Meaning that the destination is something _inside_ of the mesh like another microservice that isn't registered in the Istio Service Registry.

2. is what this issue is for and improving the visualization for. Even though the communication is pod to pod, the apps aren't talking directly to each other. The request is still going through the proxy on the sending end, and being received by the proxy on the receiving (destination_workload) end.

**Comment by nrfox** [comment]
> My initial thought is that Kiali should just omit the intermediate PassthroughCluster edges. Meaning, the graph would only show PassthroughCluster when it is a terminal node. That would result in direct edges from the source workload to the destination workload. These edges would have the same look as other edges, they would be protocol-specific.

Maybe this can go behind a display options flag that's only enabled unless explicitly checked since it's still good to know that your traffic is going through the PassthroughCluster unless you explicitly ignore it like for this issue.

**Comment by nrfox** [comment]
> If a ServiceEntry were defined I think everything would be fine

+1 this is what mesh internal service entries are for. If you've got a lot of dynamic entries like you potentially do for client side load balancing, then you'll probably need a custom way to automatically sync with the istio service registry.

**Comment by jshaughn** [comment]
> Maybe this can go behind a display options flag that's only enabled unless explicitly checked since it's still good to know that your traffic is going through the PassthroughCluster unless you want to ignore it like for this issue.

Well, I think this is an unusual case, this is the first time I've seen PassthroughCluster as an intermediate node.  I'd prefer not to add yet another option flag.  And this flag would probably need to make it to the server-side processing, which is where, I think, we'd want to prune these edges.  I'd prefer to just do it by default, at least until someone complains about it missing :-)

**Comment by MugenTwo** [comment]
>As an aside, allowing PassthroughCluster is not usually recommended for production systems, it means meshConfig.outboundTrafficPolicy.mode is set to ALLOW_ANY (the default). I think in general it is recommended to use REGISTRY_ONLY. See here for more.

For that use case that I have (which is a system composed of a couple of hundreds of services), I think starting with ALLOW_ANY is the pragmatic way to go because it allows me to do incremental changes needed to migrate my system to Istio. I think this is also the reason why ALLOW_ANY is the default option, because I think generally it is practical to work this way until reaching a point in which we can set meshConfig.outboundTrafficPolicy.mode to REGISTRY_ONLY. I am not sure if I will end up doing it, because my system doesn't really need the protection, but it is definitely something that I will consider.

I think the problem here is that it is counter-intuitive that **Istio considers a traffic that goes from a known pod directly to another known pod as an outbound traffic**, because it is not really going out of bounds does it? Istio knows which the origin was and Istio knows which the destination is. So why would it need a ServiceEntry, when there really is not entry or exit?. In fact, the more I think about it the more I think we should ask the people from Istio to reconsider and inspect this use case further. Maybe we should get ask for a more elaborate response from them?

>If a ServiceEntry were defined I think everything would be fine (note - ServiceEntry would be required if ALLOW_ANY were changed to REGISTRY_ONLY).

I did consider adding ServiceEntry for this and it would mean adding a ServiceEntry that whitelists the IP range that gets assigned to internal k8s pods (internal traffic). I think it can be a possible "patch" to contemplate as of right now (I'll check this and will let you know about the result as soon as I can), but this might not be a "solution" because there certainly are more people that will migrate their system that has "Spring Boot Admin" type of services and the more important scenario with client-side load balancing (ribbon).

>Well, I think this is an unusual case, this is the first time I've seen PassthroughCluster as an intermediate node. I'd prefer not to add yet another option flag. And this flag would probably need to make it to the server-side processing, which is where, I think, we'd want to prune these edges. I'd prefer to just do it by default, at least until someone complains about it missing :-)

The reason why I suggested to have an option to filter out non terminal nodes passthrough clusters is because some users might complain about this (as of right now, I really can't see the scenarios but it is a possibility). And I agree with you that starting with not having them until someone complains to keep things simple is a good starting point.

**Comment by jshaughn** [comment]
@MugenTwo I'm going to make the proposed change.  Would you have the ability to test a build if I were to push one to my repo, or do you need an official release to test?

**Comment by MugenTwo** [comment]
> @MugenTwo I'm going to make the proposed change. Would you have the ability to test a build if I were to push one to my repo, or do you need an official release to test?

I will check this tomorrow. Will let you know as soon as possible. Thanks and have a good day!

**Comment by MugenTwo** [comment]
We will try to test the build version. Best regards.

**Comment by jshaughn** [comment]
Hi @MugenTwo , you should be able to pull the kiali-4488 tag from https://quay.io/repository/jshaughn/kiali?tab=tags.  This build should eliminate intermediate PassthroughCluster service nodes, even when the Show Service Nodes Display option is enabled.  You should see only the direct edges from workload to workload.   Please let me know how it looks.  Also, you had mentioned seeing two edges in your CURL tests.  I think you should see only a single edge.  Assuming that is true, please ensure the request rate looks correct, and is not double what you may expect.  The change also touches code for ResponseTime and Throughput edge labels, as well as the Security Display option.  If everything else looks good, you may want to sanity check those options as well.  Let me know how it goes!

**Comment by jshaughn** [comment]
Hi @MugenTwo , any feedback on the test image?

**Comment by MugenTwo** [comment]
It will take a couple more of days before I can test it. Will let you know asap. Best regards.

**Comment by jshaughn** [comment]
@MugenTwo I would still like to hear if this is working well for you, but if you can't test an unofficial image, I have added the change to v1.44, to be released on Friday.

**Comment by MugenTwo** [comment]
I should be able to test this around 2nd/3rd week of December. Will let you know then. Thanks! Best regards.

**Comment by jshaughn** [comment]
@MugenTwo Perhaps you moved on to something else, but if you do have any feedback we'd still love to hear it.

**Comment by MugenTwo** [comment]
Hi @jshaughn ! Sorry, I haven't had time to come back to you due to other priorities. But I will make sure to provide feedback as soon as I can. Thanks again!
